### ISSUE TITLE
HMR Does not work

### ISSUE THREAD

**Issue by harrisKAsher** [module-federation/vite#183]
Hi, I am trying to migrate from @originjs/vite-plugin-federation to this as it seems better supported and has HMR support. But I cannot get the HMR to work from the remotes (though the app loads). Currently I am just trying to get this set up between 3 different apps.

This is my host/vite.config.ts
`import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { federation } from '@module-federation/vite';
import dotenv from 'dotenv';
const env = dotenv.config({ path: `../../.env` });

export default defineConfig({
	plugins: [
		federation({
			name: 'host',
			remotes: {
				core: {
					type: 'module',
					name: 'core',
					entry: `${env.parsed?.VITE_CORE_URL}/assets/remoteEntry.js`,
					entryGlobalName: 'core',
					shareScope: 'default'
				},
				globalStore: {
					type: 'module',
					name: 'globalStore',
					entry: `${env.parsed?.VITE_GLOBAL_STORE_URL}/assets/remoteEntry.js`,
					entryGlobalName: 'globalStore',
					shareScope: 'default'
				}
			},
			filename: 'assets/remoteEntry.js',
			shared: ['react', 'react-dom', 'react-router-dom']
		}),
		react()
	],
	build: {
		modulePreload: false,
		target: 'esnext',
		minify: false,
		cssCodeSplit: false
	}
});
`

This is my core/vite.config.ts

`import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { federation } from '@module-federation/vite';
import dotenv from 'dotenv';
const env = dotenv.config({ path: `../../.env` });

export default defineConfig({
	plugins: [
		federation({
			name: 'core',
			filename: 'assets/remoteEntry.js',
			remotes: {
				globalStore: {
					type: 'module',
					name: 'globalStore',
					entry: `${env.parsed?.VITE_GLOBAL_STORE_URL}/assets/remoteEntry.js`,
					entryGlobalName: 'globalStore',
					shareScope: 'default'
				}
			},
			exposes: {
				'./App': './src/App.tsx'
			},
			shared: ['react', 'react-dom', 'react-router-dom']
		}),
		react()
	],
	build: {
		modulePreload: false,
		target: 'esnext',
		minify: false,
		cssCodeSplit: false
	}
});
`

My globalStore/vite.config.ts

`import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { federation } from '@module-federation/vite';

export default defineConfig({
	plugins: [
		federation({
			name: 'globalStore',
			filename: 'assets/remoteEntry.js',
			exposes: {
				'./Store': './src/store.tsx'
			},
			shared: ['react', 'react-dom', 'react-router-dom']
		}),
		react()
	],
	build: {
		modulePreload: false,
		target: 'esnext',
		minify: false,
		cssCodeSplit: false
	}
});
`

The globalStore shouldn't matter as right now I am just trying to get the HMR working between Host and Core and have basically made my App.tsx to look like this in core. I have to keep globalStore in my config though as otherwise it will break as other components need it even though they aren't imported anywhere.

`import React, { useEffect } from 'react';

export default function App() {
	return (
		<div>Dev App Test</div>
	);
}
`

And in Host its this 

`import React from 'react';
const Core = React.lazy(() => import('core/App'));

export default function App() {
	return (
		<>
			<Core />
		</>
	);
}
`

So as you can see, just about as barebones as it gets, and yet HMR does not work. Though in the console I do see this

`[vite] hot updated: /src/App.tsx [client:223:18](http://localhost:3100/@vite/client)`

And in the network tab I can see this, which even includes the changes I made to core/App.tsx! Which just baffels me that they are there, yet it will not update them on my host.

`import { createHotContext as __vite__createHotContext } from "/@vite/client";import.meta.hot = __vite__createHotContext("/src/App.tsx");import __vite__cjsImport0_react_jsxDevRuntime from "/node_modules/.vite/deps/react_jsx-dev-runtime.js?v=f6c91154"; const jsxDEV = __vite__cjsImport0_react_jsxDevRuntime["jsxDEV"];
import RefreshRuntime from "/@react-refresh";
const inWebWorker = typeof WorkerGlobalScope !== "undefined" && self instanceof WorkerGlobalScope;
let prevRefreshReg;
let prevRefreshSig;
if (import.meta.hot && !inWebWorker) {
    if (!window.__vite_plugin_react_preamble_installed__) {
        throw new Error("@vitejs/plugin-react can't detect preamble. Something is wrong. See https://github.com/vitejs/vite-plugin-react/pull/11#discussion_r430879201");
    }
    prevRefreshReg = window.$RefreshReg$;
    prevRefreshSig = window.$RefreshSig$;
    window.$RefreshReg$ = (type, id) => {
        RefreshRuntime.register(type, "/Users/MyUserName/Projects/MyApp/apps/core/src/App.tsx " + id);
    };
    window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
}
export default function Index() {
    return /* @__PURE__ */ jsxDEV("div", { children: "Dev App Test Change" }, void 0, false, {
        fileName: "/Users/MyUserName/Projects/MyApp/apps/core/src/App.tsx",
        lineNumber: 25,
        columnNumber: 5
    }, this);
}
_c = Index;
var _c;
$RefreshReg$(_c, "Index");
if (import.meta.hot && !inWebWorker) {
    window.$RefreshReg$ = prevRefreshReg;
    window.$RefreshSig$ = prevRefreshSig;
}
if (import.meta.hot && !inWebWorker) {
    RefreshRuntime.__hmr_import(import.meta.url).then((currentExports) => {
        RefreshRuntime.registerExportsForReactRefresh("/Users/MyUserName/Projects/MyApp/apps/core/src/App.tsx", currentExports);
        import.meta.hot.accept((nextExports) => {
            if (!nextExports) return;
            const invalidateMessage = RefreshRuntime.validateRefreshBoundaryAndEnqueueUpdate("/Users/MyUserName/Projects/MyApp/apps/core/src/App.tsx", currentExports, nextExports);
            if (invalidateMessage) import.meta.hot.invalidate(invalidateMessage);
        });
    });
}
`

Either I am missing or something is broken. If you need further information from me please let me know. Thank you for the support.

**Comment by gioboa** [comment]
Hi, you can use [this](https://github.com/antfu/vite-plugin-restart) plugin to update your remotes in the host application when something is changing

**Comment by harrisKAsher** [comment]
Hi, I am not sure how that helps? I thought this package was suppose to work with HMR by default? I didn't see anything in the Documentation that I needed to use a separate plugin to get HMR. 

That being said, I am not sure what your are suggesting what I do with that plugin, would it be putting something like this in my host/vite.cofnig.ts?

```ts
ViteRestart({
			restart: [`${env.parsed?.VITE_CORE_URL}/assets/remoteEntry.js`]
		})
```

If so, that doesn't work.

**Comment by gioboa** [comment]
Yep, you need to add this plugin in the host vite configuration. It takes the file system paths to watch

**Comment by harrisKAsher** [comment]
Okay... That is a touch disappointing as I was hoping there would be a way to have the reset just happens. The way your describing would require me to have every developer edit the path so that it points to their location on their machine.  

I guess to elaborate a bit more, my hope was to have my host app allow connections from other teams and be hosted on an AWS S3 Bucket. Then we could point the Host at the remote localhost and we could develop that way so other teams didn't have to have access to the host app. But I am assuming that is not possible, correct? (Just a note, I did test that concept with @originjs/vite-plugin-federation, it works, just couldn't do HMR.)

**Comment by ScriptedAlchemy** [comment]
Im not sure if vite ever natively supported hmr, as far as i know fast refresh and hmr only work in webpack and rspack based builds as a built-in solution.

**Comment by zhangHongEn** [comment]
The built-in HMR in Vite works well. You can see it in action by starting both the host and remote in dev mode.

**Comment by zhangHongEn** [comment]
Your host is in build mode, and the remote is in dev mode. For this scenario, additional configuration is required.

**Comment by zhangHongEn** [comment]
1.	React must use the development version.
	2.	react-refresh should be a singleton.
	3.	Install React Chrome DevTools.

**Comment by ScriptedAlchemy** [comment]
Does my chrome dev tools for federation work with this package? 
Id so we have a module proxy that comes with fast refresh and hmr baked in. Allows us to hmr and develop against production with a local override

**Comment by zhangHongEn** [comment]
> Does my chrome dev tools for federation work with this package?

yes

<hr />

Can it bypass the rules of React Refresh?

**Comment by ScriptedAlchemy** [comment]
When you enable hmr switch a runtime plugin is added that uses resolve share hook and injects react runtime and dev versions from unpkg

**Comment by zhangHongEn** [comment]
> When you enable hmr switch a runtime plugin is added that uses resolve share hook and injects react runtime and dev versions from unpkg

Then it should theoretically be supported already. I’ll check the shared configuration soon and see how to adapt it to work with @vitejs/plugin-react.

@harrisKAsher Have you tried the Module Federation Chrome extension?

**Comment by zhangHongEn** [comment]
<img width="1389" alt="image" src="https://github.com/user-attachments/assets/02efb76a-930a-4a88-8261-eb71103d22e4">
<img width="1372" alt="image" src="https://github.com/user-attachments/assets/77d69bec-78aa-4a97-9138-dcf1a20f0590">

@ScriptedAlchemy  Didn’t see unpkg.

**Comment by ScriptedAlchemy** [comment]
Does a prod and dev mix work tho? Give it a try. What i know is the devtool implements a dev react to bypass mixing issues. 
Check if the chrome tools runtime plugin is added in the federation globals. It might be something we just import and add in our plugin. Not sure if extention injects it or if we auto register it on our side

**Comment by harrisKAsher** [comment]
@zhangHongEn I just installed the Module Federation Chrome extension to try it out and no matter what I do I get told "No ModuleInfo Detected".

**Comment by ScriptedAlchemy** [comment]
> @zhangHongEn I just installed the Module Federation Chrome extension to try it out and no matter what I do I get told "No ModuleInfo Detected".

Did you use the json remote protocol. Not the js one

**Comment by harrisKAsher** [comment]
No, I am using the `js` one? I tried just renaming the entry to be `remoteEntry.json` but that just results in a JSON parse error. I have also added `manifest: true` to each of my configs as the error message suggested that could be the problem. But it seems to be my files are still being generated as a js file not a json file. (I will include the error below in case it helps).

```js
Error: [ Federation Runtime ]: [ Federation Runtime ]: [ Federation Runtime ]: Failed to get manifest.
args: {"manifestUrl":"http://localhost:3110/assets/remoteEntry.json","moduleName":"globalStore"}
https://module-federation.io/guide/troubleshooting/runtime/RUNTIME-003
Original Error Message:
 SyntaxError: JSON.parse: unexpected character at line 2 column 3 of the JSON data
 ```

**Comment by ScriptedAlchemy** [comment]
The docs state you have to use the json protocol

**Comment by zhangHongEn** [comment]
@harrisKAsher mf-manifest.json
https://module-federation.io/configure/manifest.html

**Comment by ScriptedAlchemy** [comment]
As far as i know, vite doesnt support circular imports, like remote importing host exposed modules, its one-way - currently rspack/webpack implementations have omnidirectional support

Is a remote trying to import something from host?

**Comment by harrisKAsher** [comment]
No, it shouldn't be.

Host loads GlobalStore and Core
Core load GlobalStore
GlobalStore loads nothing else

When I was using @originjs/vite-plugin-federation I was able to have core load itself to get the GlobalStore, but I separated it out for this as it was clearly wasn't working, but when I just use `remoteEntry.js` without the mf-manifest.json it loads just fine.

**Comment by zhangHongEn** [comment]
@harrisKAsher  Could you provide a reproducible example? I plan to resolve the circular references and other related issues that are causing the loop.​

**Comment by zhangHongEn** [comment]
For example, I found that @module-federation/bridge-vue3 also causes Vite to call init multiple times. I need to make it more robust.

**Comment by ScriptedAlchemy** [comment]
Multiple initializations aren't a problem. It's a warning, but there's no real danger. Next.js does it too.

It's probably circular remote imports, which I'm not sure if Vite just doesn't work with cyclic host remotes or if it doesn't work with cyclic remotes at all.

**Comment by zhangHongEn** [comment]
https://github.com/module-federation/vite/blob/166052450d6b556431da32adf81452d33b756c32/src/virtualModules/virtualRemoteEntry.ts#L133

Here’s the faulty code; it leads to an infinite loop.

**Comment by ScriptedAlchemy** [comment]
Okay so in our bundlers we track if it was already initialized, so circular remotes do not infinitely initialize

**Comment by zhangHongEn** [comment]
I haven’t yet thought carefully about how to log it without polluting the global scope.

**Comment by harrisKAsher** [comment]
@zhangHongEn  Do you still need a reproducible example?

**Comment by ScriptedAlchemy** [comment]
https://github.com/module-federation/core/blob/main/packages/webpack-bundler-runtime/src/initializeSharing.ts

I also export a createContainer function from my packages which just uses webpacks actual mechanics directly for initialization. 
https://github.com/module-federation/core/blob/main/packages/webpack-bundler-runtime/src/container.ts

**Comment by zhangHongEn** [comment]
> @zhangHongEn Do you still need a reproducible example?

I haven’t reproduced the infinite loop myself, but it indeed needs optimization. No example is needed.

**Comment by harrisKAsher** [comment]
Sorry, not sure I follow, what is the init token? I can't find anything in the docs about such a thing.

**Comment by ScriptedAlchemy** [comment]
> Sorry, not sure I follow, what is the init token? I can't find anything in the docs about such a thing.

Speaking to maintainers in thread. Sorry!

**Comment by harrisKAsher** [comment]
Oh, okay. Thought it was a direct response to me. 😅

Also, another thing I am encountering (not sure its related or not) is in a remote app when I import a component from a package (ComponentA) that internally imports another component (ComponentB) from a dependent package, it fails to load the entire federated remote app. If I already import ComponentB in the remote app then everything works. This issue doesn't occur when running I go to the remote apps endpoint and view it outside the federated instance. It also doesn't happen if I do a build and preview it in the Federated Instance. I have tried a bunch of different adjustments to make ComponentA and/or ComponentB shared with the vite config but none that changes anything.

**Comment by ScriptedAlchemy** [comment]
I'm not sure about that one. 

Good litmus test is to try it with rspack or rsbuild (our vite-like experience).

If the same setup works with rspack/rsbuild, then it's a problem with vite implementation. 

If it doesn't, then it's a bug in our runtime.

**Comment by gioboa** [comment]
> Okay, well I was pretty close to done, so I just finished it up. Here it is in case it helps.
> 
> https://github.com/harrisKAsher/Federation-Vite-Demo

Thanks for sharing, I'll copy your example in the repo to show how to do it

**Comment by harrisKAsher** [comment]
> I'm not sure about that one.
> 
> Good litmus test is to try it with rspack or rsbuild (our vite-like experience).
> 
> If the same setup works with rspack/rsbuild, then it's a problem with vite implementation.
> 
> If it doesn't, then it's a bug in our runtime.

So I was trying to reproduce this and I cannot do it without one thing: getting from a private npm repository. My organization has an internal package of components to make things uniform across the various web applications we make. Its kind of similar to [HeadlessUI](https://headlessui.com/). Anyway, that is what is breaking the app, and some of them are pretty simple. Literally some of them are just div's with styles on them. But it seems that if an internal package relies on another internal package that is when it breaks.

My only theory is it is breaking because of how the federation in dev mode can't get the internal package for some reason? Not really sure how to verify that though as I am not that familiar with internal packages.  I tried with other packages  that are similar such as HeadlessUI but I could not get it to break in the same way.

**Comment by talkohavy** [comment]
I was reading intently thus far 👀 waiting to see a solution on this.

**Everything** @harrisKAsher described is happening to me as well. Line by line. Word for word.

Here's is my example, which is very much similar to his: [vite-micro-frontends-react](https://github.com/talkohavy/vite-micro-frontends-react). Like him, I kept things very basic & simple (although my case has 1 parent host & 2 child remotes, no grand-children, 1 child is on build (appears on host's "remotes"), and the other child is dynamically imported).

**Comment by zhangHongEn** [comment]
I will prioritize solving this problem as the top priority when I have time

---- Replied Message ----
| From | ***@***.***> |
| Date | 11/19/2024 07:20 |
| To | module-federation/vite ***@***.***> |
| Cc | zhn ***@***.***>,
Mention ***@***.***> |
| Subject | Re: [module-federation/vite] HMR Does not work (Issue #183) |

I was reading intently thus far 👀 waiting to see a solution on this.

***@***.*** described is happening to me as well. Line by line. Word for word.

Here's is my example, which is very much similar to his: vite-micro-frontends-react. Like him, I kept things very basic & simple (although my case has 1 parent host & 2 child remotes, no grand-children, 1 child is on build (appears on host's "remotes"), and the other child is dynamically imported).

**Comment by ScriptedAlchemy** [comment]
So for me to support. I'll require a rspack or webpack based attempt. Then i can help. Otherwise we will need to rely on our community partners who maintain vite to proxy the needs back to me from a runtime perspective. 

Core team only allocated resources to 
- rspack users
- webpack users
- vanilla runtime users 
Vite plugin implements vanilla runtime, but then request has to come from user (vite maintainer/implementer) 

We can wait for @zhangHongEn to take a look and contact Core team, but just letting yall know where my support kicks in. 🥰🥰

**Comment by ScriptedAlchemy** [comment]
Is it like missing css or is there something else like missing JS modules. Missing css would be tailwind tree shake classes it can't see. That can be solved with allow list. If it's missing modules on the other hand. That's something we need to investigate etc

**Comment by zhangHongEn** [comment]
This issue mentions two problems:

1. Infinite loop
2. React HMR (Hot Module Replacement)

**Comment by ScriptedAlchemy** [comment]
Just a heads up. You can contact me on Twitter / bluesky 

I have a LLC dedicated to NDA signing operations for such cases. Would allow you to send me a zip of your node modules and all that I can just extract and run without any liabilities on either end. 

Pain in the ass but we can pull that never IF it's needed. On my end it's not a problem. Sign a few hundred a year haha 😄

**Comment by zhangHongEn** [comment]
Thanks to @ScriptedAlchemy  for providing a solution approach, which allows me to implement it without checking the pack code; the only thing missing is time.
1: Solution is <a href="#issuecomment-2477441832">here</a>
2: I haven’t seen the Chrome plugin successfully injecting React

**Comment by ScriptedAlchemy** [comment]
> Thanks to @ScriptedAlchemy  for providing a solution approach, which allows me to implement it without checking the pack code; the only thing missing is time.
> 
> 1: Solution is <a href="#issuecomment-2477441832">here</a>
> 2: I haven’t seen the Chrome plugin successfully injecting React

Ahhh time. The great evil of all mankind 🫠

**Comment by gioboa** [comment]
Thanks @ScriptedAlchemy for your core help. this means a lot for this project.
Can I suggest to use the official [MF public channel](https://discord.com/channels/1055442562959290389/1055447490628956161)
Using this channel we avoid lack of alignment.

**Comment by talkohavy** [comment]
Alright guys, here's the news.
I'll start from the end:

HMR _does_ work for `rsbuild` ✅
HMR _doesn't_ work for `@module-federation/vite`... ❌

I've created a small project that illustrate the 2 bugs: one is the infinite loop, and the other is the HMR not working.

The project: [module-federation-with-vite-vs-rsbuild](https://github.com/talkohavy/vite-micro-frontends-react)

Steps to reproduce:

## HMR works with Rsbuild

1. checkout from `master` to a branch called `rsbuild`
2. Navigate to `apps/booksMF`, and run:

```bash
pnpm install
pnpm rsdev
```

3. Navigate to `apps/host`, and run:

```bash
pnpm install
pnpm rsdev
```

4. In your IDE, open file `booksMF/src/components/App/App.tsx`,  and change the text of the `h1` from "Vite + React" to whatever, and see that the **HMR worked properly**.

---

## HMR doesn't works with Vite

5. Stop the rsbuild of both `host` and `books` servers.
6. Navigate to `apps/booksMF`, and run:

```bash
pnpm dev
```

7. Navigate to `apps/host`, and run:

```bash
pnpm dev
```

The result: **an infinite loop!**

## Different Approach (where vite's wepabb works but HMR doesn't):

8. Open up the `vite.config.ts` of booksMF, and replace `manifest: true,` (line 15) with `filename: 'remoteEntry.js',` (line 15). Uncomment one and comment out the other. Now, in `vite.config.ts` of host, change `entry: 'http://localhost:3001/mf-manifest.json',` to `entry: 'http://localhost:3001/remoteEntry.js',`.

Result: try running both projects now with `pnpm dev` - you'll succeed! But if you'l' open up the file `booksMF/src/components/App/App.tsx`, and change something in it, you'll fond that the **HMR doesn't work**.

**Comment by harrisKAsher** [comment]
> Is it like missing css or is there something else like missing JS modules. Missing css would be tailwind tree shake classes it can't see. That can be solved with allow list. If it's missing modules on the other hand. That's something we need to investigate etc

I am assuming this in response to my importing issue from a private repository for packages? I have realized thats probably a separate issue so I have made a new issue [here](https://github.com/module-federation/vite/issues/187). But just for the record, there is missing css, but that's not the problem as the components my organization makes uses styled components. So I believe it is missing the module.

**Comment by jamalsoueidan** [comment]
@harrisKAsher did you get the HMR to work in the host with the example? or it doesnt work?

**Comment by harrisKAsher** [comment]
@jamalsoueidan In Host if you comment out the GlobalStore and Core App code then the host app will run and update with HMR. But keep in mind you have to run it in dev mode, and dev mode breaks with GlobalStore or Core App imported right now.

**Comment by harrisKAsher** [comment]
Oh, I was talking about my working example. Not sure what to tell you about the example they have.

https://github.com/harrisKAsher/Federation-Vite-Demo

**Comment by jamalsoueidan** [comment]
> Oh, I was talking about my reproduction of the bug example. Not sure what to tell you about the example they have.
> 
> https://github.com/harrisKAsher/Federation-Vite-Demo

I want to find a working example of host and remote with HMR working, just simple example, the example they have, work with CSS, but not with TSX. It seems like their is something not correct with shared config in vite, because the host is getting app.tsx network update everytime I change the file in the remote.

**Comment by DanielSegal1** [comment]
Are there plans to fix it anytime soon?
I tried to fiddle with it and found out that when changing a file in the remote, it gets updated in the sources of chrome
but only in that file. files that are dependent on it, still import the old version of the file.

**Comment by gioboa** [comment]
I'm little bit lost here. I read the thread few times but I don't understand if there is a valid solution here or if we need to implement something form scratch

**Comment by jamalsoueidan** [comment]
> I'm little bit lost here. I read the thread few times but I don't understand if there is a valid solution here or if we need to implement something form scratch

RSbuild works, other setup didn't.

**Comment by codelzb** [comment]
@gioboa hello, I don't understand the data changes in vue, but the content of the template page doesn't change. Why is it related to hmr? In my understanding, hmr is a development-stage feature used to automatically update the page when the code is changed, without completely refreshing the browser.

**Comment by protoEvangelion** [comment]
@zhangHongEn I know you were taking a look at this last year. Any plans to pick this back up. If not, any pointers for the community and how we can carry this forward?

**Comment by protoEvangelion** [comment]
@ScriptedAlchemy @harrisKAsher @zhangHongEn I have a workaround which fixes HMR for host/remote on the same react tree. Basically, you just add this to vite config `plugins`. All it does is replace the import to react-refresh, ensuring the port is the host port:

```
export function getSingleReactRefreshPlugin(): PluginOption {
  return {
    name: 'single-react-refresh', // Name for the plugin
    enforce: 'pre', // Run this plugin before other transformations
    transform(code, id) {
      if (/\.(js|ts|jsx|tsx)$/.test(id)) {
        const updatedCode = code.replace(
          /import RefreshRuntime from "\/@react-refresh";/g,
          'import RefreshRuntime from "http://localhost:4000/t1/@react-refresh";'
        )
        return updatedCode
      }
      return null // Return null if no transformation is applied
    }
  }
}
```

I tried solving this with `singleton` but still two versions of react-refresh were getting loaded & HMR then doesn't work. I presume that is because the import that gets injected into each file by vite-plugin-react is the forward slash: `/@react-refresh` which means it doesn't get pulled directly from node_modules.

I'm not confident that this is the most elegant or speedy solution, so my question is, is there theoretically a better alternative? If not, could we bake it into this plugin?

**Comment by harrisKAsher** [comment]
@protoEvangelion I tried this out in my project and can't get it to work. I assume this is only needed in the remote apps? I am wondering if something else you did that I am not doing.

**Comment by protoEvangelion** [comment]
@harrisKAsher Correct it's only needed in remotes. Are you using `@vitejs/plugin-react`? That is the package responsible for injecting react-refresh. If you inspect one of the js files in your network panel you should see something like:

```
import RefreshRuntime from "/@react-refresh";
```

So all my plugin fix does is replace this with the absolute url. You can verify it's working if you only see one "react-refresh" loaded in the network: 

![Image](https://github.com/user-attachments/assets/afe6dd8b-5284-446d-9680-30f4c75be75e)

I debugged the whole HMR flow & it works as expected. Here is the simplified flow:


1. save some remote file
2. vite devServer chokidar picks up change
3. vite devServer notifies @vite/client
4. @vite/client re imports the file via dynamic import
5. @vite/client calls callbacks subscribed to by react-refresh via vite api: `import.meta.hot.accept`
6. react-refresh applies the updates to the react tree using  [scheduleRefresh](https://github.com/facebook/react/blob/4632e36a4ef16a1af24987c56e42b664f6403e64/packages/react-reconciler/src/ReactFiberHotReloading.js#L222C14-L222C29)

The missing piece is that without the text replace plugin, there are two instances of react-refresh & thus the remote react-refresh does apply updates to the main react tree.

**Comment by protoEvangelion** [comment]
@harrisKAsher do you have a demo branch I can pull?

**Comment by harrisKAsher** [comment]
No, I can't create a demo. My application is pretty complex, I tried reproducing it the simple demo I have shared before, but your solution works great there. And yes `react()` needs to be above it for it to work it seems.

My real world application has multiple layers to it, so I will have to dig in and see which layer is causing the HMR with this work around to break. I will post here if I figure out my apps issues.

---

I also thought I would take time to mention something I mentioned long ago I wanted to try and have tried in case someone wonders if this work and finds these. Using Module Federation with a Remote Local Host. This does in fact work! But there is one caveat, you have to have the main "host" app running locally.


So the set up is this: 
In your actual host app that is "released" have its urls pointing at the `localhost` URLs of your remote apps.

In your project that has remote apps you will have your remote apps as well as a sudo app that I just called "host" this app doesn't get released and is only used locally. The "host" app should basically just module federate in the actual host app that is released. Then that actual host app module federates in your local app in and you can work on your remote app with out running the host app on your machine.

The reason for the caveat is that without the fake local host app and just pointing at the live app, it is running in build mode, so it is missing stuff such as HMR that an in dev mode would expect and it creates a problem that causes the remote apps to fail to load.

**Comment by jamalsoueidan** [comment]
@harrisKAsher  if you have time maybe you can create a simpel repos example you can share, thank you :)

**Comment by harrisKAsher** [comment]
Here is a reproduction:

https://github.com/harrisKAsher/Federation-Vite-Demo/tree/signle-refresh-plugin

Check out the packages/other components. I have 3 Components, ComponentA, ComponentB, and ComponentC. They all import the globalStore which is module federated in in the core app. However, since all 3 of them are using it HMR doesn't work. Comment out 1 of them in any of the components, refresh your page, and HMR starts working again.

**Comment by gioboa** [comment]
@harrisKAsher I cloned your repo but I can't reproduce the issue. I can't find ComponentA etc. etc.
Can you check if this PR #269 solves your case? Thanks

**Comment by harrisKAsher** [comment]
@gioboa Sorry, I don't have a ton of time to look into this. I think I got it installed (never have installed from a Github PR before) and it's not working.

You can't just pull the repo, you have to be on the branch `signle-refresh-plugin` that is the one with ComponentA, etc. If you don't see that when you run the app your on the wrong branch.

**Comment by gioboa** [comment]
@harrisKAsher #269 is not solving HMR in your scenario

**Comment by protoEvangelion** [comment]
@harrisKAsher HMR does work in this repo with the plugin I mentioned. The reason it isn't working in your case is because you have @demo/app installed like an npm package instead of just importing or federating it. See the demo below:

https://github.com/user-attachments/assets/cc311f08-329a-4e43-862b-2fbc0981072c

@gioboa what we need to decide on is which direction do we want to go. I like the idea of #269 but it doesn't fix the problem because the `react()` plugin still injects `/react-refresh` into every file. I see two options:

1. We can add a new option in this vite plugin like `hostPort` & add the replace plugin I mentioned above
2. Try to get a PR open on the react refresh plugin to support adding a prefix

I think #2 would be ideal but #1 would be faster.

**Comment by gioboa** [comment]
I see, the #2 I think is the right option to implement.
On the other hand, I'm wondering about the official plugin that is cooking [here](https://github.com/rolldown/rolldown/tree/main/examples%2Fmodule-federation)

cc @underfin can you tell me more about the progress on the official implementation please?

**Comment by harrisKAsher** [comment]
@protoEvangelion Correct, because that is how turborepo works. You install the items in the packages as a package into your other apps. It isn't meant to be federated in or imported with a bunch of `../`. It is treated like a package.

That being said if it is an edge case; don't let that hold everyone else up on getting a fix for this.

**Comment by protoEvangelion** [comment]
#2 it is @gioboa. Once this is merged we can update docs and close this issue. https://github.com/vitejs/vite-plugin-react/pull/420

@harrisKAsher Makes sense but this is a limitation with vite itself as it does not pick up changes in node_modules folder. There are some workarounds though: https://stackoverflow.com/questions/74992987/vite-how-to-change-code-in-node-modules-folder-to-debug-find-errors

**Comment by protoEvangelion** [comment]
> wondering about the official plugin that is cooking [here](https://github.com/rolldown/rolldown/tree/main/examples%2Fmodule-federation)

@underfin @gioboa correct me if I'm wrong...this vite mf plugin will still be necessary for those who have muti-pack systems webpack + vite + rspack (_like myself_) because the rolldown impl is not based on the module-federation runtime package. As nice as rolldown mf looks, I don't think I could ever use it unless all our build systems were migrated to vite.

https://github.com/rolldown/rolldown/blob/bc1eba75e7046fda952ad3001e7b6c9a2f2b1200/crates/rolldown_plugin_module_federation/src/lib.rs#L41

**Comment by underfin** [comment]
> I see, the [#2](https://github.com/module-federation/vite/issues/2) I think is the right option to implement. On the other hand, I'm wondering about the official plugin that is cooking [here](https://github.com/rolldown/rolldown/tree/main/examples%2Fmodule-federation)
> 
> cc [@underfin](https://github.com/underfin) can you tell me more about the progress on the official implementation please?

> because the rolldown impl is not based on the module-federation runtime package

@protoEvangelion @gioboa The rolldown builtin module federation plugin is implemented base on module federation 2.0, it could be a rust port of the plugin for rolldown, but it only work at build, it will be recommend usage at rolldown-vite full bundle module. The main feature is finished, the dts/hmr related features need to implement , we are researching at now.

**Comment by gioboa** [comment]
@underfin Can you share an example of the official rolldown-vite module federation please?

**Comment by underfin** [comment]
> [@underfin](https://github.com/underfin) Can you share an example of the official rolldown-vite module federation please?

https://github.com/rolldown/rolldown-vite-module-federation-example you could see at here.

**Comment by gioboa** [comment]
> you could see at here.

That's amazing 👏👏

**Comment by protoEvangelion** [comment]
@gioboa We are good to close this issue out now because HMR is fixed [here](https://github.com/vitejs/vite-plugin-react/pull/420) & released in vite-plugin-react@4.4.0

@harrisKAsher's other issue regarding HMR is a limitation by design in vite itself to not watch for file changes in node_modules, so I don't think HMR will ever work for that case.

**Comment by gioboa** [comment]
Thanks @protoEvangelion for your amazing help 👏
